collector/agentPoller.go

```diff
@@ -15,451 +15,433 @@
+	agentMetrics, err := a.getAgentMetrics()
 	if err != nil {
-		return nil, err
+		log.Errorf("Failed to get agent metrics from %s:%d. Error: %s", a.AgentIP, a.Port, err)
+		return metrics
 	}
 
-	json, err := rootJSON.Object()
+	log.Debugf("Fetching container metrics from agent %s:%d", a.AgentIP, a.Port)
+	containerMetrics, err := a.getContainerMetrics()
 	if err != nil {
-		stats <- statsd.MakeEvent(statsd.AgentQueryBadData)
-		return nil, err
+		log.Errorf("Failed to get container metrics from %s:%d. Error: %s", a.AgentIP, a.Port, err)
+		return metrics
 	}
 
-	nowMillis := time.Now().UnixNano() / 1000000
-	// collect datapoints
-	// expecting a single dict containing 'string => floatval' entries
-	var datapoints []interface{}
-	for key, valRaw := range json.Map() {
-		valFloat, err := valRaw.Float64()
-		if err != nil {
-			stats <- statsd.MakeEvent(statsd.AgentMetricsValueUnsupported)
-			log.Printf("Failed to convert value %s to float64: %+v", key, valRaw)
-			continue
-		}
-		datapoint, err := goavro.NewRecord(datapointNamespace, datapointSchema)
-		if err != nil {
-			log.Fatalf("Failed to create Datapoint record for topic %s (agent %s): %s",
-				a.Topic, agentState.agentID, err)
-		}
-		datapoint.Set("name", systemMetricPrefix+strings.Replace(key, "/", ".", -1)) // "key/path" => "key.path"
-		datapoint.Set("time_ms", nowMillis)
-		datapoint.Set("value", valFloat)
-		datapoints = append(datapoints, datapoint)
-	}
-	stats <- statsd.MakeEventCount(statsd.AgentMetricsValue, len(datapoints))
-	if len(datapoints) == 0 {
-		return nil, errors.New("No datapoints found in agent metrics")
-	}
+	log.Debugf("Finished polling agent %s:%d, took %f seconds.", a.AgentIP, a.Port, time.Since(now).Seconds())
 
-	metricListRec, err := goavro.NewRecord(metricListNamespace, metricListSchema)
-	if err != nil {
-		log.Fatalf("Failed to create MetricList record for topic %s (agent %s): %s",
-			a.Topic, agentState.agentID, err)
-	}
-	metricListRec.Set("topic", a.Topic)
-	metricListRec.Set("datapoints", datapoints)
-	// note: agent_id tag is automatically added downstream
-	metricListRec.Set("tags", make([]interface{}, 0))
-	// just use a size of zero, relative to limits it'll be insignificant anyway:
-	return &AvroDatum{metricListRec, a.Topic, 0}, nil
+	metrics.agentState = agentState
+	metrics.agentMetricsSnapshot = agentMetrics
+	metrics.containerMetrics = containerMetrics
+	metrics.timestamp = now.Unix()
+
+	return metrics
 }
 
-// fetches container state from the agent (via /state) to populate AgentState
-func (a *Agent) getAgentState(stats chan<- statsd.StatsEvent) (*AgentState, error) {
-	rootJSON, err := a.getJSONFromAgent("/state", agentTestStateFileFlag, stats)
-	if err != nil {
-		return nil, err
+// transform will take metrics retrieved from the agent and perform any
+// transformations necessary to make the data fit the output expected by
+// producers.MetricsMessage.
+func (a *Agent) transform(in metricsMeta) (out []producers.MetricsMessage) {
```

Comment:
This method is a bit larger than I would normally care for, and there is some duplicated code between "produce agent metrics" and "produce container metrics." This can probably be extracted into a helper function that builds a new `producers.MetricsMessage{}` and returns, and might make testing a bit easier. :thinking: